### PR TITLE
Integration

### PR DESCRIPTION
**Pull request by makoto**
## PRs to merge

- https://github.com/wearekickback/contracts/pull/43
- https://github.com/wearekickback/contracts/pull/45
- https://github.com/wearekickback/contracts/pull/52

## Description

- https://github.com/wearekickback/contracts/issues/41
- https://github.com/wearekickback/contracts/issues/42

## List of features added/changed

- `sendAndWitdraw`
- `clearAndSend `
 
## Screenshots (if appropriate):

Testing on our local env. The video walkthrough is https://www.youtube.com/watch?v=MuWHu7UQRD0 (we haven't built UI to call clearAndsend yet)

## Checklist:

- [ x ] My code follows the code style of this project.
- [ x ] My code implements all the required features.

### CODE REVIEW

File: contracts/AbstractConference.sol [hunk 1]
```diff
@@ -85,18 +108,44 @@ contract AbstractConference is Conference, GroupAdmin {
...
 
+    /**
+    * @dev sendAndWithDraw function allows to split _payoutAmount_ among addresses in _addresses_.
+    * 
+    * _addresses_ contains ethereum addresses
+    * _values_ contains the value of eth/dai to give to addresses
+    * 
+    * addresses[i] will receive values[i]
+    * The function emits the event SendAndWithdraw with the following informations:
+    * (addresses, values, participant address, payoutAmount - sum(values), payoutAmount)
```
Comment (makoto) [inline contracts/AbstractConference.sol:126]: Add an additional comment mentioning that the  leftover money  is  sent  to ` msg.sender`

File: contracts/AbstractConference.sol [hunk 2]
```diff
@@ -85,18 +108,44 @@ contract AbstractConference is Conference, GroupAdmin {
...
+        require(addresses.length == values.length, 'more addresses than values or viceversa');
+
+        participants[msg.sender].paid = true;
+
+        uint256 sumOfValues = 0;
+        for(uint i = 0; i < addresses.length; i++) {
+            sumOfValues = sumOfValues.add(values[i]);
+            
+            if(sumOfValues > payoutAmount)
+                revert('payout amount is less than sum of values');
```
Comment (makoto) [inline contracts/AbstractConference.sol]: Can re rewritten as follows

```
require(sumOfValues <= payoutAmount, 'payout amount is less than sum of values');
```

File: contracts/Conference.sol [hunk 3]
```diff
@@ -10,6 +10,7 @@ interface Conference {
     event RegisterEvent(address addr, uint256 index);
     event FinalizeEvent(uint256[] maps, uint256 payout, uint256 endedAt);
     event WithdrawEvent(address addr, uint256 payout);
+    event SendAndWithdraw(address payable[] addresses, uint256[] values, address addr, uint256 payoutLeft);
```
Comment (makoto) [inline contracts/Conference.sol]: Rename to `SendAndWithdrawEvent`

File: contracts/AbstractConference.sol [hunk 4]
```diff
@@ -85,18 +108,44 @@ contract AbstractConference is Conference, GroupAdmin {
...
-        participant.paid = true;
+    function withdraw() external onlyEnded canWithdraw {
+        participants[msg.sender].paid = true;
+        withdrawn = withdrawn.add(1);
         doWithdraw(msg.sender, payoutAmount);
         emit WithdrawEvent(msg.sender, payoutAmount);
     }
 
+    /**
+    * @dev sendAndWithDraw function allows to split _payoutAmount_ among addresses in _addresses_.
```
Comment (makoto) [inline contracts/AbstractConference.sol]: `sendAndWithDraw` -> `sendAndWithdraw`

File: test/behaviors/conference.behavior.js [hunk 5]
```diff
@@ -466,6 +465,71 @@ function shouldBehaveLikeConference () {
...
+      let previousBalance = await getBalance(registered);
+
+      await conference.sendAndWithdraw([], [], {from:registered});
+      
+      let diff = (await getBalance(registered)).sub(previousBalance);
+      assert.isOk(diff.gt( mulBN(deposit, 0.9) ))
+
+      assertBalanceWithDeposit((await getBalance(conference.address)), mulBN(deposit, 1))
+    })
+
```
Comment (makoto) [inline test/behaviors/conference.behavior.js:470]: Add additional test to donate everything

File: test/behaviors/conference.behavior.js [hunk 6]
```diff
@@ -466,6 +465,71 @@ function shouldBehaveLikeConference () {
...
+      await register({conference, deposit, user:owner, owner});
+      await register({conference, deposit, user:registered, owner});
+
+      assertBalanceWithDeposit((await getBalance(conference.address)), mulBN(deposit, 2))
+    })
+
+    it('should take only `deposit` value', async function() {
+      await conference.cancel({from:owner});
+      await conference.sendAndWithdraw([accounts[2]], [donation], {from:registered});
+      assertBalanceWithDeposit((await getBalance(conference.address)), mulBN(deposit, 1))
```
Comment (makoto) [inline test/behaviors/conference.behavior.js:438]: Add an assertion to make sure that the remainder is sent to the attendee

File: test/behaviors/conference.behavior.js [hunk 7]
```diff
@@ -466,6 +465,71 @@ function shouldBehaveLikeConference () {
...
+      let previousBalanceTwo = await getBalance(accounts[20]);
+
+      await conference.sendAndWithdraw([accounts[10], accounts[20]], [donation, donationTwo], {from:registered});
+      
+      let diffOne = (await getBalance(accounts[10])).sub(previousBalanceOne);
+      let diffTwo = (await getBalance(accounts[20])).sub(previousBalanceTwo);
+      
+      assert.isOk(diffOne.eq(donation));
+      assert.isOk(diffTwo.eq(donationTwo));
+
```
Comment (makoto) [inline test/behaviors/conference.behavior.js:486]: Add an assertion to make sure that the remainder is sent to the attendee

File: contracts/AbstractConference.sol [hunk 8]
```diff
@@ -208,20 +276,55 @@ contract AbstractConference is Conference, GroupAdmin {
...
+
+        uint256 fee = payoutAmount.mul(clearFee).div(1000);
+        uint256 toAttenders = payoutAmount.sub(fee);
+
+        uint256 totalSent = 0;
+        for(uint256 j = lastSent.div(256); totalSent < _num && j < attendanceMaps.length; j++) {
+            uint256 map = attendanceMaps[j];
+            for(uint256 i = 0; totalSent < _num && i < 256; i++) {
+                Participant storage participant = participants[participantsIndex[(j.mul(256).add(i).add(1))]];
+                if(0 < (map & (2 ** i)) && !participant.paid) {
```
Comment (makoto) [inline contracts/AbstractConference.sol:299]: Add more comment

File: contracts/Conference.sol [hunk 9]
```diff
@@ -47,12 +48,15 @@ interface Conference {
     // AbstractConference
     function register() external payable;
     function withdraw() external;
+    function sendAndWithdraw(address payable[] calldata, uint256[] calldata) external;
     function totalBalance() view external returns (uint256);
     function isRegistered(address _addr) view external returns (bool);
     function isAttended(address _addr) external view returns (bool);
```
Comment (makoto) [inline contracts/Conference.sol:56]: Add more comment on the implementation of isAttended

File: test/behaviors/conference.behavior.js [hunk 10]
```diff
@@ -513,26 +577,195 @@ function shouldBehaveLikeConference () {
       assertBalanceWithDeposit((await getBalance(conference.address)), mulBN(deposit, 1))
     })
 
-    it('owner receives the remaining if cooling period is passed', async function(){
-      conference = await createConference({coolingPeriod:0});
-      deposit = await conference.deposit();
+    it('owner receives the remaining if cooling period is passed and everyone called withdraw', async function(){
```
Comment (makoto) [inline test/behaviors/conference.behavior.js:541]: Consider removing the clear altogether as the dust will be extremely small

File: test/behaviors/conference.behavior.js [hunk 11]
```diff
@@ -513,26 +577,195 @@ function shouldBehaveLikeConference () {
...
+
+      let previousBalance = (await getBalance(owner))
+
+      let conferenceBalance = (await getBalance(conference.address))
+      assert.isOk(conferenceBalance.gt(0))
+
+      await conference.clear({from:owner})
+
+      let diff = previousBalance.sub((await getBalance(owner)));
+      assert.isOk(conferenceBalance.eq(conferenceBalance));
```
Comment (makoto) [inline test/behaviors/conference.behavior.js]: Change to  assert.isOk(diff.eq(conferenceBalance));

File: test/behaviors/conference.behavior.js [hunk 12]
```diff
@@ -513,26 +577,195 @@ function shouldBehaveLikeConference () {
...
+
+      await wait(20, 1);
+      await conference.clearAndSend(1);
+      
+      let diffOne = new EthVal(await getBalance(accounts[11])).sub(previousBalanceOne)
+      let diffTwo = new EthVal(await getBalance(accounts[12])).sub(previousBalanceTwo)
+      
+      assert.isOk(diffOne.eq(payoutAmount.sub(fees)))
+      assert.isOk(diffTwo.lt(payoutAmount.sub(fees)))
+
```
Comment (makoto) [inline test/behaviors/conference.behavior.js:624]: Add assertion that diffTwo  should  be  zero

**Comment by makoto** [comment]
General test Consideration

- Add check revert message using ` truffle-assertions or` or ` @openzeppelin/test-helpers`

File: contracts/AbstractConference.sol [hunk 13]
```diff
@@ -85,18 +108,44 @@ contract AbstractConference is Conference, GroupAdmin {
...
+        for(uint i = 0; i < addresses.length; i++) {
+            sumOfValues = sumOfValues.add(values[i]);
+            
+            if(sumOfValues > payoutAmount)
+                revert('payout amount is less than sum of values');
+
+            doWithdraw(addresses[i], values[i]);
+        }
+                
+        uint256 amountLeft = payoutAmount.sub(sumOfValues);
```
Comment (makoto) [inline contracts/AbstractConference.sol:142]: Add `withdrawn = withdrawn.add(totalSent);`

File: contracts/AbstractConference.sol [hunk 14]
```diff
@@ -197,8 +265,8 @@ contract AbstractConference is Conference, GroupAdmin {
      * @param _maps The attendance status of participants represented by uint256 values.
      */
     function finalize(uint256[] calldata _maps) external onlyAdmin onlyActive {
-        uint256 totalBits = _maps.length * 256;
-        require(totalBits >= registered && totalBits - registered < 256, 'incorrect no. of bitmaps provided');
+        uint256 totalBits = _maps.length.mul(256);
+        require(totalBits >= registered && totalBits.sub(registered) < 256, 'incorrect no. of bitmaps provided');
```
Comment (makoto) [inline contracts/AbstractConference.sol]: `totalBits >= registered` is unnecessary as `totalBits.sub(registered) < 256` does the same check.

The modified one will be as follows.

```require(totalBits.sub(registered) < 256, 'incorrect no. of bitmaps provided');```

Comment (hiddentao) [reply contracts/AbstractConference.sol]: This is true

File: contracts/AbstractConference.sol (hunk 8 above)
Comment (hiddentao) [reply contracts/AbstractConference.sol:299]: `Check to see if bit number "i" is set`

File: test/behaviors/conference.behavior.js (hunk 10 above)
Comment (makoto) [reply test/behaviors/conference.behavior.js:541]: We will keep this as dust could increase once we introduce promotional return

**Pull request merged**